### PR TITLE
Fix dynamicFee and transaction group encoding

### LINKED ISSUE #108: Transaction Group support for ASA Transfers
The _'get_obj_for_encoding'_ method on the Transaction class does not seem to set the _'grp'_ parameter for ASA transfers. This causes a transaction group to fail with the error message _'TransactionPool.Remember: transactionGroup: [0] had zero Group but was submitted in a group of 2'_.

I had to monkey patch the method to get it to work,
```
for (let mp of txns) {
    let orig = mp.get_obj_for_encoding;
    let replace = function() {
        let obj = orig.bind(this)();
        obj.grp = this.group;
        return obj;
    };
    mp.get_obj_for_encoding = replace.bind(mp);
}
```

### LINKED ISSUE #110: Fix dynamic fee contract

### PR DESCRIPTION
**Pull request by EvanJRichard**
## Summary
This PR fixes dynamicFee, which was producing incorrect `stxn`s. While getting that working, a bug relating to transaction groups was fixed.

### Testing
Golden testing.

### See also
This fixes #108 but not #100 
This closes #110

### CODE REVIEW

File: src/logicTemplates/dynamicfee.js [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+ * @param {dict} txn - main transaction from payer's signDynamicFee output (a dict of constructor arguments, NOT a transaction.Transaction)
+ * @param {LogicSig} lsig - the signed logic received from the payer's signDynamicFee output
+ * @param {Uint8Array} privateKey - the private key for the account that pays the fee
+ * @param {int} fee - fee per byte for both transactions
+ * @param {int} firstValid - first protocol round on which both transactions will be valid
+ * @param {int} lastValid - last protocol round on which both transactions will be valid
+ *
+ * @throws on invalid lsig
+ */
+function getDynamicFeeTransactions (txn, lsig, privateKey, fee, firstValid, lastValid) {
```
Comment (winder) [inline src/logicTemplates/dynamicfee.js]: Should this be `getReimbursementTransaction`? Maybe I named the function wrong in java.

File: src/logicTemplates/dynamicfee.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    if (!lsig.verify(address.decode(txn.from).publicKey)) {
+        throw new Error("invalid signature");
+    }
+
+    txn.firstRound = firstValid;
+    txn.lastRound = lastValid;
+    txn.fee = fee;
```
Comment (winder) [inline src/logicTemplates/dynamicfee.js]: This is fee per byte, so you'll need `txn.fee = fee * estimatedSize(txn)`

File: src/logicTemplates/dynamicfee.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    if (txn.fee < transaction.ALGORAND_MIN_TX_FEE) {
+        txn.fee = transaction.ALGORAND_MIN_TX_FEE
+    }
+
+    let keys = nacl.keyPairFromSecretKey(privateKey);
+    let from = address.encode(keys.publicKey);
+
+    // must remove lease and re-add using addLease so that fee calculation will match other SDKs
```
Comment (winder) [inline src/logicTemplates/dynamicfee.js]: You should be able to remove all the `txnObj` stuff and just use `txn` after fixing the calculation above.

File: src/logicTemplates/dynamicfee.js (hunk 1 above)
Comment (EvanJRichard) [reply src/logicTemplates/dynamicfee.js]: could be either, I think both are accurate, submitting the resultant bytes both sends a transaction with a dynamic fee and a transaction reimbursing the fee

File: src/logicTemplates/dynamicfee.js (hunk 2 above)
Comment (EvanJRichard) [reply src/logicTemplates/dynamicfee.js]: `estimateSize` is a method on an already-constructed transaction, so letting the constructor handle the feePerByte conversion seems to be the only way to end up with the same fee as in the other SDKs.

Another example of doing this: a transaction can be constructed with `lease` from the outset, but using `addLease` is necessary to replicate the same fee output.

File: src/logicTemplates/dynamicfee.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        if (!Number.isSafeInteger(firstValid) || firstValid < 0) throw Error("firstValid must be a positive number and smaller than 2^53-1");
+        if (!Number.isSafeInteger(lastValid) || lastValid < 0) throw Error("lastValid must be a positive number and smaller than 2^53-1");
+
+        if (closeRemainder === undefined) {
+            closeRemainder = this.zeroAddress
+        }
+        if (lease === undefined) {
+            let leaseBytes = nacl.randomBytes(32);
+            lease = Buffer.from(leaseBytes).toString('base64');
+        }
```
Comment (btoll) [inline src/logicTemplates/dynamicfee.js]: Do you mean to explicitly test for only `undefined` here?  Could you do `if (!closeRemainder)` or even set a default value in the constructor signature (where it's readable, of course)?

These are nits :)

File: src/logicTemplates/dynamicfee.js [hunk 5]
```diff
@@ -0,0 +1,180 @@
+const templates = require('./templates');
```
Comment (pzbitskiy) [inline src/logicTemplates/dynamicfee.js]: We already have this file in master, what is the difference?

**Pull request merged**